core/types/celo_transaction_marshalling.go

```diff
@@ -30,20 +30,71 @@ func celoTransactionMarshal(tx *Transaction) ([]byte, bool, error) {
 	enc.Hash = tx.Hash()
 	enc.Type = hexutil.Uint64(tx.Type())
 	switch itx := tx.inner.(type) {
+	case *LegacyTx:
```

Comment:
How does this relate to `LegacyTx` handling code in `MarshalJSON`? Can we just have it in one place?

Reply:
This is the same as the LegacyTx handling code in MarshalJSON, but with the extra fields for celo legacy transactions. 

The approach I'm taking here is to try to segregate the celo transaction handling from the upstream, so that maintaining our fork is easier. We could just add our handling in the transaction_marshalling.go file, but then we'd be creating a highly interleaved diff which would likely make updating from upstream more difficult.

It would be nice if we could call MarshalJSON from here to get the upstream marshalling and then simply add the celo fields, but because MarshalJSON first calls celoTransactionMarshal we'd overflow the stack.

So it seemed a reasonable tradeoff to me to duplicate the unmarshalling code for legacy transactions here in order to simplify fork maintenance.

Let me know if you think there is a bette way to do this?

Reply:
Yeah, makes sense to separate that. Now we just need to make sure that we change our code when the upstream code changes.

Reply:
Well, I'm hoping it wont change much since we are just duplicating the legacy transaction. There's only been one functional change on that code since 2021.